[Dart 3] Records (#4766)
dart-lang/site-www: src/_guides/libraries/library-tour.md

print('Done with all the long steps!');
```

`Future.wait()` returns a future which completes once all the provided
futures have completed. It completes either with their results,
or with an error if any of the provided futures fail. 

#### Handling errors for multiple futures

You can also wait for parallel operations on an [iterable]({{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-async/FutureIterable/wait.html)
or [record]({{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-async/FutureRecord2/wait.html)
of futures.

These extensions return a future with the resulting values of all provided
futures. Unlike `Future.wait`, they also let you handle errors. 

If any future in the collection completes with an error, `wait` completes with a
[`ParallelWaitError`][]. This allows the caller to handle individual errors and
dispose successful results if necessary.

> Use `wait` on an _iterable_ of futures when you _don't_ need the result values
> from each individual future:

Reviewer: **suggestion:** Condition, then instruction per [GSG](https://developers.google.com/style/sentence-structure).
```suggestion
When you _don't_ need the result values from each individual future,
use `wait` on an _iterable_ of futures.
```

print('Done with all the long steps!');
```

`Future.wait()` returns a future which will complete once all the provided
futures have completed, either with their results, or with an error if any of
the provided futures fail. 

#### Handling errors for multiple futures

You can also wait for parallel operations on an [iterable]({{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-async/FutureIterable/wait.html)
or [record]({{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-async/FutureRecord2/wait.html)
of futures.

These extensions return a future with the resulting values of the provided futures.
Unlike `Future.wait()`, if any future in the collection completes with an error,
`wait` completes with a [`ParallelWaitError`][] that allows the caller to
handle individual errors and dispose successful results if necessary:

```dart
void main() async {
  Future<bool> delete() async =>  ...
  Future<bool> copy() async =>  ...
  Future<bool> causesError() async =>  ...

  try {

>     await [delete(), copy(), causesError()].wait;    // List of futures

Reviewer 2: This code is not using the result value. 
It's the first/only example of using `await iterable.wait`, so I think it should show a *full* use-case.

It might be the only example of using the `.wait` getter some people will see.

Consider:

  bool success;
  try {
    /// Wait for each future in list.
    var results = await [delete(), copy(), causesError()].wait;
    success = !results.contains(false);
  } on ParallelWaitError<List<bool?>, List<AsyncError?>> catch (e) {
    success = false;
    // print ... all the same prints.
  }
  if (success) print("Success");
```

Author: @lrhn Thanks for fixing this! Is the `success` part necessary? I realize it's adding extra clarification that the `try` block represents successfully waiting for all the futures, and that `ParallelWaitError` represents a failure, despite having results returned, but isn't that kind of implied by the nature of `try` anyway?

Author: Ok, I think I understand now. I'll change them to `Future<void>` so the example is at least realistic (though still simpler, imo). 

If you think the use case shown in the example you wrote is much more valuable, though, let me know and I'll use it. I just worry that readers might conflate some of the other stuff that's going on in that example (like using `contains()` or somewhat "unpacking" the results into a boolean variable) with the question of "how to use `wait` on an iterable".

Reviewer 2: I think switching to `Future<void>` is fine. 
I expect "just tell me when they're done" to be the most common use-case for list-wait.

Having an _unknown_ number of same-typed futures, that you want to run in parallel, and still want all the values of, seems more like an edge case for a badly-written API than a common use-case.
(If you end up doing `var results = await ids.map((id) => server.fetch(id)).wait;`, I'd ask why the server API doesn't provide a `server.fetchMultiple(Iterable<Id>)` function to begin with.)

It'll still happen. A lot. But I have no doubt people doing that will already know `Future.wait`.

    Expect.fail("Didn't throw");
  } on ParallelWaitError<List<bool?>, List<AsyncError?>> catch (e) {   
    print(e.values[0]);    // Prints result of future
    print(e.values[1]);    // Prints result of future
    print(e.values[2]);    // Prints null

    print(e.errors[0]);    // Prints null
    print(e.errors[1]);    // Prints null
    print(e.errors[2]);    // Preints error
  }
}
```

Waiting on a record of futures enables the same functionality, with the additional 
feature that the futures can be of different types:

```dart
void main() async {
  Future<int>    delete() async =>  ...
  Future<String> copy() async =>  ...

>   Future<bool>   causesError() async =>  ...

Author: I like it too, but I'll skip it for now in favor of a wider discussion

  try {

>     await (delete(), copy(), causesError()).wait;    // Record of futures

Reviewer 2: Again, consider capturing the values, even if it's just:
```dart
  (int, String, bool) result = await (delete(), copy(), errorResult()).wait;
```

The fact that record-`.wait` allows different types of results 
is it's biggest feature, so it bears highlighting.

Author: Similar to my last question, is the `Expect.fail()` line right below necessary? I know I added this part, but now I'd like to get anything that's not absolutely necessary out for brevity. 

I was hoping the `errorResult()` future would imply that the result is expected to fail, just for the sake of the example. And I'm hoping if they want to learn more they'll go to the API docs for more complete examples (assuming that [the test samples](https://dart-review.googlesource.com/c/sdk/+/288903/12/tests/lib/async/future_record_extension_test.dart) show up in the api docs somewhere, and if they don't, they really should!)

Reviewer 2: This is example code, possibly the first/only example code someone sees of `await (someRecord).wait`.

Unlike awaiting a list, the only reason to await a record of futures is if you *need* the result values. Otherwise you could just have made a `List<Future<void>>` instead of a record.

So I think the first example someone sees of `await (someRecord).wait` should show someone who *wants* to use the result values.

But that comes back to the purpose of the example:
* Is it to show an example of `await (someRecord).wait`, or
* Is it to show how `await (someRecord).wait` differs from ... something? `Future.wait`?

The latter seems too context dependent. If someone reads this example first, they won't have the context needed. They will read it as an example of using `await (someRecord).wait`, so it should be a *good* example of that first, and then something which can also be compared to other features second.

IMO. :)

Author: That's a _really_ good point. I was thinking that it'd be most useful to users, for some reason, to just see the bare minimum of how to set up using `wait` on a record. I didn't think of showing an actual useful use case. Thanks so much for explaining!

Author: > the only reason to await a record of futures is if you need the result values.

I added this description on top of the example, and showed some arbitrary "use" of the result values

print('Done with all the long steps!');
```

`Future.wait()` returns a future which will complete once all the provided
futures have completed, either with their results, or with an error if any of
the provided futures fail. 

#### Handling errors for multiple futures

You can also wait for parallel operations on an [iterable]({{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-async/FutureIterable/wait.html)

> or [record]({{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-async/FutureRecord2/wait.html)

Reviewer 3: ```suggestion
of [record]({{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-async/FutureRecord2/wait.html)
```

Right?

Author: Records aren't actually part of the iterable family. So these future APIs for iterables and records are separate. 

Kind of unrelated, but the implementation distinction between records' implementation / API and the rest of the collections API has made it hard to generalize all them with a single word (e.g. "you can aggregate different values in a single object using Dart's collection types (lists, maps, sets) ... _and_ records"). I wish I could just call them a collection.

Reviewer 2: The word I'd use is that they're *composite* objects (can contain other objects).
More precisely "composite value containers" (even if it's a bit redundant).

Most objects are composite in some sense, but some objects has that as their primary objective, being exactly the sum of their parts with no extra semantic interpretation or functionality. A list is just a list. A task queue might be a list, but it has a role. It's not _just_ a data container.

"Collections" are traditionally *mutable* homogeneous composite objects, whose primary task is to collect and access their values. Very traditionally they are lists, sets, queues and/or maps (maps are sometimes considered a collection, other times not, qv., `java.util.Collection` and `java.util.Map`). Good old "data structures".

"Structs", "records" or "tuples" are heterogeneous composite objects, but still mostly without extra functionality on top of just carrying the data values. Aka. "data objects"/"data classes".

Don't know if that helps at all :grin:

Author: Composite! That's exactly what I was looking for. I've weaseled my way around lumping records in with the rest of the collections so far, so I don't think I need to go back and correct anything, but if the need arises again (and I'm sure it will) I'll use "composite". Might even be a good entry for the new glossary. Thanks!

print('Done with all the long steps!');
```

`Future.wait()` returns a future which will complete once all the provided
futures have completed, either with their results, or with an error if any of

> the provided futures fail. 

Author: Worked around to shorten this, because a future completing is different than returning (I don't understand it much more than that...):

> `Future.wait()` returns a future which completes once all the provided
futures have completed. It completes either with their results,
or with an error if any of the provided futures fail.

#### Handling errors for multiple futures

You can also wait for parallel operations on an [iterable]({{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-async/FutureIterable/wait.html)
or [record]({{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-async/FutureRecord2/wait.html)
of futures.

> These extensions return a future with the resulting values of all the provided
> futures. Unlike `Future.wait`, if any future in the collection completes with
> an error, `wait` completes with a [`ParallelWaitError`][] that allows the caller
> to handle individual errors and dispose successful results if necessary.

Reviewer: **issue:** Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading).
```suggestion
These extensions return a future with the values of all provided futures.
This differs from `Future.wait`.
If any future in the collection returns an error, `wait` returns a [`ParallelWaitError`][].
The caller can then handle each error and drop successful results if needed.
```

Author: I went through your reviews bottom up on this section, but now that I'm up here, I realized that I could kind of rearrange the whole section to be more concise. I noticed I repeated what the new `wait` does 3 times (how it differs from `Future.wait`, then how it works for iterables, then how it works for records, which are all the same description). So I changed everything to now only mention how it works _once_ at the top, and then focus only on the differences:

> These extensions return a future with the resulting values of all provided
> futures. Unlike `Future.wait`, they also let you handle errors. 
> 
> If any future in the collection completes with an error, `wait` completes with a
> [`ParallelWaitError`][]. This allows the caller to handle individual errors and
> dispose successful results if necessary.
(NOTE: I think "completes" is different than "returns", so had to leave that in)
> 
> Use `wait` on an _iterable_ of futures when you _don't_ need the result values
> from each individual future:
> 
> ```dart
> // examples the same
> ```
> 
> Use `wait` on a _record_ of futures when you _do_ need the individual result values
> from each future. This also provides the additional benefit that the futures can be of different types:
>
> ```dart
> // examples the same
> ```
> (end section)

}
```

> Use `wait` on a _record_ of futures when you _do_ need the individual result values
> from each future. This also provides the additional benefit that the futures can be of different types:

Reviewer: **suggestion:** Condition, then instruction per [GSG](https://developers.google.com/style/sentence-structure).
```suggestion
When you _need_ the individual result values from each future,
use `wait` on a _record_ of futures.
This also allows the futures can be of different types.
```

[NativeFinalizer]: {{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-ffi/NativeFinalizer-class.html
[NoSuchMethodError]: {{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-core/NoSuchMethodError-class.html
[Object]: {{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-core/Object-class.html

> [`ParallelWaitError`]: {{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-async/ParallelWaitError-class.html

Reviewer 4: why does this one use backticks when the others don't?

Author: I'd argue that the class names that _aren't_ back ticked need to change?
